File: 0945. Minimum Increment to Make Array Unique/minimum_increment_to_make_array_unique.py
from typing import List


class Solution:
    def minIncrementForUnique_MK1(self, A: List[int]) -> int:
        A.sort()
        ret = 0
        for i in range(1, len(A)):
            if A[i] <= A[i - 1]:
                ret += A[i - 1] + 1 - A[i]
                A[i] = A[i - 1] + 1
        return ret

    def minIncrementForUnique_MK2(self, A: List[int]) -> int:
        buff = [0] * 80001
        for n in A:
            buff[n] += 1
        ret, size = 0, len(A)
        for i in range(len(buff) - 1):
            if buff[i] == 1:
                size -= 1
            elif buff[i] > 1:
                buff[i + 1] += buff[i] - 1
                ret += buff[i] - 1
                size -= 1
            if not size:
                break
        return ret

File: 0945. Minimum Increment to Make Array Unique/minimum_increment_to_make_array_unique.cpp
#include <algorithm>
#include <vector>
using namespace std;

class Solution
{
public:
    int minIncrementForUnique_MK1(vector<int> &A);
    int minIncrementForUnique_MK2(vector<int> &A);
};

int Solution::minIncrementForUnique_MK1(vector<int> &A)
{
    vector<int> buff(80001, 0);
    for (const int &n : A)
        buff[n]++;
    
    int ret = 0;
    int len = A.size();
    for (size_t i = 0; i < buff.size() - 1 && len; i++) {
        if (buff[i] > 1) {
            ret += buff[i] - 1;
            buff[i + 1] += buff[i] - 1;
            len--;
        }
        else if (buff[i] == 1)
            len--;
    }
    return ret;
}

int Solution::minIncrementForUnique_MK2(vector<int> &A)
{
    sort(A.begin(), A.end());
    int ret = 0;
    for (size_t i = 1; i < A.size(); i++) {
        if (A[i] <= A[i - 1]) {
            ret += A[i - 1] + 1 - A[i];
            A[i] = A[i - 1] + 1;
        }
    }
    return ret;
}
